File: CPP/12_Ransom_Note.cpp
/*
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.
*/

bool canConstruct(string ransomNote, string magazine) {
    if(magazine.size() < ransomNote.size()) {
        return false;
    }
    unordered_map<char, int> note_counter;
    for(int i = 0; i < magazine.size(); i++) {
            note_counter[magazine[i]]++;
    }
    for(int i = 0; i < ransomNote.size(); i++) {
        if(note_counter[ransomNote[i]] > 0) {
            note_counter[ransomNote[i]]--;
        } else {
            return false;
        }
    }
    return true;
}

File: Python/12_Ransom_Note.py
'''
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.
'''

def canConstruct(ransomNote, magazine):
    note_counter = {}
    for c in magazine:
        if c not in note_counter:
            note_counter[c] = 1
        else:
            note_counter[c] += 1
    for c in ransomNote:
        if c not in note_counter:
            return False
        if note_counter[c] > 0:
            note_counter[c] -=1
        else:
            return False
    return True